man/tcos-server-utils.1
.TH tcos-server-utils 1 "Sep 24, 2007" "tcos-server-utils man page"

.SH NAME
tcos-server-utils \- Python script to exec some actions on connected thin clients.

.SH DESCRIPTION

.PP
\fBtcos-server-utils\fP is a python script that generate XMLRPC events to do some jobs:
 * Send a message
 * Reboot
 * Poweroff
 * Restart Xorg

tcos-server-utils use ~/.tcosconfig.conf auth and need to be configured before using it.


.SH OPTIONS
.B [no options]
 With no options show some actions.

.B --help
 Some usage help about tcos-server-utils

.B --action=foo
 Exec desired action ( reboot, poweroff, restartx, message, nothing )
   (action nothing is a self test)

.B --text="foo"
 If action=message this will the text displayed

.B --users=user1,user2
 Coma separated list of usernames to send text message

.B --debug
 Show verbose output when running

.SH BUGS
Contact Mario Izquierdo <mariodebian@gmail.com>

Homepage: http://www.tcosproject.org

.SH AUTHOR
Mario Izquierdo (mariodebian) <mariodebian@gmail.com>


Changelog
tcosmonitor (0.1.18.1) unstable; urgency=low

  * Move dbus conf file and tcos-dbus-client to tcos-tcosmonitor

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Sat, 07 Jul 2007 14:02:41 +0200

tcosmonitor (0.1.18.0) unstable; urgency=low

  * Remove pyexamples dir
  * xmlrpc/sh/restartx.sh:
    - call restartxorg instead of killing X
  * tcos-devices-ng:
    - Allot to disable floppy cdroms or hdd from conf file
      (disable_floppy=1, disable_cdroms=1, disable_hdd=1)
  * xmlrpc/*:
    - Add tcos.who method to retrieve username connected at :0

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Sat, 07 Jul 2007 13:28:15 +0200

tcosmonitor (0.1.17.2) unstable; urgency=low

  * xmlrpc/Makefile xmlrpc/var/etc/abyss.conf:
    - Set correct paths for tcos-standalone and tcos thin client

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Wed, 27 Jun 2007 23:54:22 +0200

tcosmonitor (0.1.17.1) unstable; urgency=low

  * New package tcos-standalone (Use tcosmonitor with not standalone hosts)
  * xmlrpc/*.h:
    - Add TCOS_PATH define to tcosxmlrpc, to allow running external tools from other paths
  * xmlrpc/sh/*:
    - Ad conditions to exec or not some tools.
  * TcosTrayIcon.py:
    - Remove StatusIcon declaration (break at gtk < 2.10)
  * Add support of standalone mode in TcoaActions.py
  * TcosActions.py:
    - patch to start/stop video broadcast (vidal_joshur at gva dot es)
    - patch to send files with rsync (vidal_joshur at gva dot es)

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Wed, 27 Jun 2007 20:10:18 +0200

tcosmonitor (0.1.16.2) unstable; urgency=low

  * udev/tcos-udevd.sh:
    - Force udev to save ID_FS_TYPE in disk, not only partitions

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Wed, 20 Jun 2007 21:18:12 +0200

tcosmonitor (0.1.16.1) unstable; urgency=low

  * xmlrpc/sh/devicesctl.sh:
    - Add support to mount NTFS with ntfs-3g driver
  * tcos-devices-ng.py:
    - Mount ntfs-3g in a thread (fuse mounts exec a process)
  * TcosXmlRpc.py:
    - Don't fail in GetDevicesInfo (use try--except)

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Wed, 20 Jun 2007 15:14:16 +0200

tcosmonitor (0.1.16) unstable; urgency=low

  * xmlrpc/sh/devicesctl.sh:
    - Fix typo in include header
  * images/*cdrom*
    - new image based on gnome theme
  * tcos-devices-ng.py:
    - Fix translation strings (vars go outside _('%s')%var )
    - Add umount_all function called from exit
    - Fix USB device remount
    - Allow to specify mounts points with a new conf file ~/.tcos-devices-ng.conf
  * Add new man page: man/tcos-devices-ng.1

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Tue, 19 Jun 2007 18:34:18 +0200

tcosmonitor (0.1.15.5) unstable; urgency=low

  * xmlrpc/sh/devicesctl.sh:
    - Option gethdd now returns all type disk except swap or iso9660

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Tue, 19 Jun 2007 12:40:51 +0200

tcosmonitor (0.1.15.4) unstable; urgency=low

  * xmlrpc/sh/devicesctl.sh:
    - Add --gethdd param and return ext3 or vfat devices in /etc/fstab
  * tcos-devices-ng.py:
    - Read hdd device partitions in thin client and put in menu
  * images/hdd_*mount.png:
    - Add hdd icons
  * po/*:
    - Update translations (new hdd strings)

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Mon, 18 Jun 2007 18:52:00 +0200

tcosmonitor (0.1.15.3) unstable; urgency=low

  * lockscreen/* :
    - Change copyright header
    - Compile with imlib-config flags and libs
  * dbus/81tcos-utils:
    - Start tcos-devices --daemon and tcos-devices --hidden
  * tcos-devices-ng.py:
    - Don't use shared vars before importing module shared

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Mon, 18 Jun 2007 11:06:10 +0200

tcosmonitor (0.1.15.2) unstable; urgency=low

  * conf/tcos.conf man/* :
    - fix some english typos (close ticket #3) thanks to Andres Jimenez again
  * lockscreen/*:
    - Add support to put a png/jpg, xpm, etc... image in lockscreen
    - Default image locked.png
  * tcos-devices-ng.py:
    - Clean code and remote extra comments
  * pyexamples:
    - Add some python examples

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Sun, 17 Jun 2007 17:01:00 +0200

tcosmonitor (0.1.15.1) unstable; urgency=low

  * shared.py:
    - Remove some ugly debug info
  * xmlrpc/sh/devicesctl.sh:
    - Add --exists parameter to know if a file (or device) exists
  * xmlrpc/sh/soundctl.sh:
    - Exis if not soundcard found
  * xmlrpc/xauth.c:
    - Compare local hostname with cookie hostname.
  * tcos-devices.py:
    - Exit if X cookie don't match
  * images/*:
    - Add some tcos-devices-ng.py images
  * TcosTrayIcon.py:
    - New Tcos class (used to build systray menu)
  * tcos-devices-ng.py:
    - New and more powerfull remote device manager
  * Rename dbus/81tcos-dbus-client to dbus/81tcos-utils and start tcos-devices from there
  * po/*:
    - Update spanish translation

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Thu, 14 Jun 2007 21:15:15 +0200

tcosmonitor (0.1.14.4) unstable; urgency=low

  * TcosXauth.py:
    - Read DISPLAY info and search for hostname and IP address

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Sat,  9 Jun 2007 19:33:04 +0200

tcosmonitor (0.1.14.3) unstable; urgency=low

  * tcos-devices.py, tcos-volume-manager.py:
    - If gtk version is greater than 2.10 use gtk.statusicon instead of egg.trayicon

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Sat, 09 Jun 2007 17:11:33 +0200

tcosmonitor (0.1.14.2) unstable; urgency=low

  * TcosXauth.py:
    - Read xauth cookie with -n param to not resolve hostname
    - Try (with try: except: to get hostname with socket) not xauth locks

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Fri, 08 Jun 2007 22:15:23 +0200

tcosmonitor (0.1.14.1) unstable; urgency=low

  * Initialize.py:
    - Fix AboutDialog, close and destry button for gtk > 2.10

server-utils/tcos-server-utils.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-
##########################################################################
# TcosMonitor writen by MarioDebian <mariodebian@gmail.com>
#
#    tcos-server-utils version __VERSION__
#
# Copyright (c) 2006 Mario Izquierdo <mariodebian@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2, or (at your option)
# any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
###########################################################################



import os, sys
import gobject
import getopt
from gettext import gettext as _



if os.geteuid() != 0:
    print "tcos-server-utils ERROR: you must be root to run this"
    sys.exit(1)



if not os.path.isfile("shared.py"):
        sys.path.append('/usr/share/tcosmonitor')
else:
        sys.path.append('./')

import shared
def print_debug(txt):
    if shared.debug:
        print "%s::%s" %("tcos-server-utils", txt)



shared.from_cron=False
actions=["reboot", "poweroff", "restartx", "message", "nothing"]
action = ""
text=""
users=""

def usage():
    print "tcos-server-utils help:"
    print ""
    print "   tcos-server-utils --action=XXX  (action must be: %s ) "                %(", ".join(actions) )
    print "                                   (use --action=nothing to test with doing nothing!!!)" 
    print "   tcos-server-utils --text=\"foo\" (if action=message this text will be displayed "
    print "                                     in all connected users with zenity)"
    print "                     --users=foo,bar (coma separated list of users we want to do action"
    print "                                      only valid for --action=message)"
    print "   tcos-server-utils -d [--debug]  (write debug data to stdout)"
    print "   tcos-server-utils -h [--help]   (this help)"


try:
    opts, args = getopt.getopt(sys.argv[1:], ":hd", ["help", "cron", "debug", "action=", "text=", "users="])
except getopt.error, msg:
    print msg
    print "for command line options use tcos-server-utils --help"
    sys.exit(2)



# process options
for o, a in opts:
    if o in ("--help"):
        usage()
        sys.exit(0)
        
    if o in ("-d", "--debug"):
        shared.debug = True
    if o in ("--cron"):
        shared.from_cron=True
    if o == "--action":
        if not a in actions:
            print "TCOS tcos-server-utils ERROR: action \"%s\" not avalaible" %(a)
            sys.exit(1)
        action = a
    if o == "--text":
        text=a
    if o == "--users":
        users=a


if action=="":
    print "TCOS tcos-server-utils ERROR: action must be in: %s" %(", ".join(actions) )
    sys.exit(0)


class ServerUtils:
    def __init__(self):
        self.name="ServerUtils"
        self.worker_running = False
        
        # get all devices
        import TcosXmlRpc
        import TcosConf
        self.config=TcosConf.TcosConf(self)
        self.xmlrpc=TcosXmlRpc.TcosXmlRpc(self)
        
        
        if self.config.GetVar("xmlrpc_username") == "" or self.config.GetVar("xmlrpc_password") == "":
            print "TCOS tcos-server-utils ERROR: need to create /root/.tcosmonitor.conf with user and pass."
            print "                         see /usr/share/doc/tcosmonitor/README.tcos-server-utils"
            sys.exit(1)
        
        import LocalData
        self.localdata=LocalData.LocalData(self)
        
        self.allclients=self.localdata.GetAllClients("netstat")
        self.alltcosclients=[]
        
        for host in self.allclients:
            self.xmlrpc.newhost(host)
            if self.xmlrpc.connected:
                print_debug ("Host %s connected" %(host) )
                self.alltcosclients.append(host)
            else:
                print_debug ("Host %s NOT connected" %(host) )
                
        if len(self.alltcosclients) == 0:
            print "TCOS tcos-server-utils No host connected, exiting..."
            sys.exit(0)
        
        print ("Doing action \"%s\" in %s" %(action, self.alltcosclients) )
        
        if action == "message":
            if text != "":
                print ( "Searching for all connected users..." )
                
                connected_users=[]
                #connected_users=['magna26']
                if users != "":
                    connected_users = users.split(',')
                    print "Users from cdmline: %s" %(connected_users)
                else:
                    
                    for client in self.allclients:
                        if self.localdata.IsLogged(client):
                            connected_users.append(self.localdata.GetUsername(client))
                
                    print ( "Connected users: %s" %connected_users)
                        
                from TcosDBus import TcosDBusAction
                self.dbus_action=TcosDBusAction( \
                    admin=self.config.GetVar("xmlrpc_username"), \
                    passwd=self.config.GetVar("xmlrpc_password") )
                
                result=self.dbus_action.do_message ( connected_users, text )
                if not result:
                    print "ERROR: sending dbus msg: %s" %(self.dbus_action.get_error_msg() )
                else:
                    print "DBus message send ok."
                    
            else:
                print ( "ERROR: message action need a --text value" )
        
        
        for client in self.alltcosclients:
            if action == "reboot":
                print ( "Restarting %s..."  %client )
                self.xmlrpc.newhost(client)
                self.xmlrpc.Exe("reboot")
            elif action == "poweroff":
                print ( "Shutting down %s..."  %client )
                self.xmlrpc.newhost(client)
                self.xmlrpc.Exe("poweroff")
            elif action == "restartx":
                print ( "Restarting Xorg of %s..."  %client )
                self.xmlrpc.newhost(client)
                self.xmlrpc.Exe("restartx")
            elif action == "message":
                pass
            else:
                print ( "Unknow action %s in client %s" %(action, client) )
                


if __name__ == "__main__":
    app=ServerUtils()

